### PR TITLE
cmd/geth: dump config for metrics

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
closes https://github.com/ethereum/go-ethereum/issues/22007

```
./geth --metrics --metrics.influxdb  dumpconfig
```

### CODE REVIEW

**Comment by MariusVanDerWijden** [comment]
@lightclient @MysticRyuujin could you take a look if that works for you?

**Comment by lightclient** [comment]
Was able to test this with the metrics configuration I am using and it works great. Thanks @MariusVanDerWijden!

**Comment by holiman** [comment]
@MariusVanDerWijden is this still WIP, or is it ready for merge? If the latter, please change the title

**Comment by holiman** [comment]
I don't get it. This does dump out the metrics config to toml ... but it doesn't read the metrics config _from_ the toml... ?

**Comment by holiman** [comment]
I would expect these two commands to give the exact same output: 
```
[user@work go-ethereum]$ ./build/bin/geth --metrics --metrics.influxdb --metrics.port=7070  --metrics.expensive dumpconfig 2>/dev/null | grep Metrics -A10
[Metrics]
Enabled = true
EnabledExpensive = true
HTTP = "127.0.0.1"
Port = 7070
EnableInfluxDB = true
InfluxDBEndpoint = "http://localhost:8086"
InfluxDBDatabase = "geth"
InfluxDBUsername = "test"
InfluxDBPassword = "test"
InfluxDBTags = "host=localhost"
[user@work go-ethereum]$ ./build/bin/geth --metrics --metrics.influxdb --metrics.port=7070  --metrics.expensive dumpconfig  2>/dev/null  > test.conf && ./build/bin/geth --config test.conf dumpconfig 2>/dev/null | grep Metrics -A10
```

**Comment by MariusVanDerWijden** [comment]
@holiman It did read the configs from toml, but it did overwrite them every time^^ It works now

**Comment by holiman** [comment]
Almost there now, but CLI should take precedence, so this should spit out `7171`, not `7070`: 
```
./build/bin/geth --metrics --metrics.port=7070 dumpconfig  2>/dev/null  > test.conf && ./build/bin/geth --metrics.port=7171 --config test.conf dumpconfig 2>/dev/null | grep Metrics -A10 | grep Port
Port = 7070
```

**Comment by MariusVanDerWijden** [comment]
Hmm it takes precedence if you enable `--metrics` explicitly
```
./build/bin/geth --metrics --metrics.port=7070 dumpconfig  2>/dev/null  > test.conf && ./build/bin/geth --metrics --metrics.port=7171 --config test.conf dumpconfig 2>/dev/null | grep Metrics -A10 | grep Port
Port = 7171
```
But I'll change it

**Comment by MariusVanDerWijden** [comment]
So the code is quite a bit worse now but I think I need it to convey the whole logic.
* If `--metrics` are enabled it should read the default values
* If the metrics are read from file, they should be overwritten by the flags

**Comment by holiman** [comment]
I don't quite understand why you give special treatment to the `enabled` flags. If

**Comment by MariusVanDerWijden** [comment]
No you're right, I thought that we wouldn't want things like influxdbpassword written in the config if it was disabled, but you're right that it should be written out anyway. changed now.
Now the metric section is always dumped even if `--metrics` or `--influxdb` are disabled.

File: cmd/geth/config.go [hunk 1]
```diff
@@ -115,12 +129,28 @@ func defaultNodeConfig() node.Config {
 	return cfg
 }
 
+func defaultMetricConfig(ctx *cli.Context) metricConfig {
```
Comment (holiman) [inline cmd/geth/config.go:132]: Perhaps this should be `metricsFromCliArgs`.. When I saw `defaultMetricConfig` being used before, I intuitively though that it just initialized some defaults, and didn't grokk that it actually read the cli arguments

**Review by holiman — COMMENTED** [review]
One minor comment, but generally looks ok. Will test it a bit

**Review by holiman — APPROVED** [review]
Works for me!

File: cmd/geth/config.go (hunk 1 above)
Comment (MariusVanDerWijden) [reply cmd/geth/config.go:132]: Yeah, that's the thing with this strange logic, we need to read the metrics twice. 
First we need to read the "default" metrics and apply them s.th. if we don't override them in the load config, they are set to the default values.
Then, after reading the config, we need to read the metrics again (this time only the specially set metrics) to overwrite the config file configs with the configs from cli.

So basically we do:
- read config from cli regardless of whether they are set (reads defaults)
- read config from file (overwrites the default configs if applicable)
- read only the set configs from cli (overwrites the file configs)

I'll change the naming!

**Review by DGKSK8LIFE — APPROVED** [review]

File: cmd/geth/config.go [hunk 2]
```diff
@@ -115,12 +129,28 @@ func defaultNodeConfig() node.Config {
...
+
 // makeConfigNode loads geth configuration and creates a blank node instance.
 func makeConfigNode(ctx *cli.Context) (*node.Node, gethConfig) {
 	// Load defaults.
 	cfg := gethConfig{
-		Eth:  eth.DefaultConfig,
-		Node: defaultNodeConfig(),
+		Eth:     eth.DefaultConfig,
+		Node:    defaultNodeConfig(),
+		Metrics: metricsFromCliArgs(ctx),
```
Comment (fjl) [inline cmd/geth/config.go:153]: I don't understand why you need this function (`metricsFromCliArgs`) if you also have `applyMetricConfig`. The idea behind the config file loading is:

The config is always created using the Go code defaults (on line 150).
Then, the TOML file is loaded to override some things.
Finally, the flags are applied to override more things.

In this PR, for metrics, it takes the defaults from flags, then overrides with config file, then applies the flags again.
I think we should remove metricsFromCliArgs and just define the defaults for metrics directly in the code.

Comment (MariusVanDerWijden) [reply cmd/geth/config.go:153]: Yeah, the problem with this approach is that we have now two different places where the default value of the flags is defined.
Once here and once in the cli.Flags.

Comment (fjl) [reply cmd/geth/config.go:153]: The way we usually solve this is by taking the flag default from the Go struct default. 

https://github.com/ethereum/go-ethereum/blob/master/cmd/utils/flags.go#L136

Comment (MariusVanDerWijden) [reply cmd/geth/config.go]: I created a new default var in the `metrics` package
